src/main/java/org/sagebionetworks/web/client/widget/doi/CreateOrUpdateDoiModal.java

```diff
@@ -0,0 +1,215 @@
+import org.sagebionetworks.repo.model.asynch.AsynchronousResponseBody;
+import org.sagebionetworks.repo.model.doi.v2.Doi;
+import org.sagebionetworks.repo.model.doi.v2.DoiCreator;
+import org.sagebionetworks.repo.model.doi.v2.DoiRequest;
+import org.sagebionetworks.repo.model.doi.v2.DoiResourceType;
+import org.sagebionetworks.repo.model.doi.v2.DoiResourceTypeGeneral;
+import org.sagebionetworks.repo.model.doi.v2.DoiTitle;
+import org.sagebionetworks.web.client.PopupUtilsView;
+import org.sagebionetworks.web.client.SynapseJavascriptClient;
+import org.sagebionetworks.web.client.events.EntityUpdatedEvent;
+import org.sagebionetworks.web.client.events.EntityUpdatedHandler;
+import org.sagebionetworks.web.client.widget.asynch.AsynchronousProgressHandler;
+import org.sagebionetworks.web.client.widget.asynch.JobTrackingWidget;
+import org.sagebionetworks.web.client.widget.entity.controller.SynapseAlert;
+import org.sagebionetworks.web.shared.asynch.AsynchType;
+
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.gwt.user.client.ui.Widget;
+import com.google.inject.Inject;
+
+public class CreateOrUpdateDoiModal implements CreateOrUpdateDoiModalView.Presenter {
+	public static final String DOI_CREATED_MESSAGE = "DOI successfully updated for ";
+	public static final String DOI_MODAL_TITLE = "Create or Update a DOI";
+
+	private CreateOrUpdateDoiModalView view;
+	private JobTrackingWidget jobTrackingWidget;
+	private SynapseJavascriptClient javascriptClient;
+	private EntityUpdatedHandler entityUpdatedHandler;
+	private Doi doi;
+	private SynapseAlert synapseAlert;
+	private PopupUtilsView popupUtilsView;
+
+	@Inject
+	public CreateOrUpdateDoiModal(CreateOrUpdateDoiModalView view,
+								  JobTrackingWidget jobTrackingWidget,
+								  SynapseJavascriptClient javascriptClient,
+								  SynapseAlert synapseAlert,
+								  PopupUtilsView popupUtilsView) {
+		this.view = view;
+		this.jobTrackingWidget = jobTrackingWidget;
+		this.javascriptClient = javascriptClient;
+		this.synapseAlert = synapseAlert;
+		this.popupUtilsView = popupUtilsView;
+		view.setSynAlert(synapseAlert);
+		view.setJobTrackingWidget(jobTrackingWidget);
+		view.setPresenter(this);
+		view.setModalTitle(DOI_MODAL_TITLE);
+	}
+
+	public void configureAndShow(String objectId, ObjectType objectType, Long versionNumber, EntityUpdatedHandler entityUpdatedHandler) {
+		view.reset();
+		synapseAlert.clear();
+		this.entityUpdatedHandler = entityUpdatedHandler;
+		doi = new Doi();
+		getExistingDoi(objectId, objectType, versionNumber);
+		view.show();
+	}
+
+	public void getExistingDoi(String objectId, ObjectType objectType, Long objectVersion) {
+		javascriptClient.getDoi(objectId, objectType, objectVersion).addCallback(new FutureCallback<Doi>() {
+			@Override
+			public void onSuccess(@NullableDecl Doi doi) {
+				populateForms(doi);
+			}
+
+			@Override
+			public void onFailure(Throwable t) {
+				doi.setObjectId(objectId);
```

Comment:
if exception is Notfound, then do this and show().  Otherwise, use popupUtils to pop up the associated error message.

Reply:
suggest constructing the local doi object if not found only.